File: packages/client/src/game/Game.ts
import { Entity } from '../ecs/Entity'
import { Manager } from '../ecs/manager'
import { Gui } from './gui/Gui'
import { State } from './State'
import { AbilitiesSystem } from './systems/Abilities'
import { AggressionSystem } from './systems/Aggression'
import { AnimateSpriteSystem } from './systems/AnimateSprite'
import { AttackSystem } from './systems/Attack'
import { CheckpointSystem } from './systems/Checkpoint'
import { ContainerSystem } from './systems/Container'
import { DamageSystem } from './systems/Damage'
import { DashSystem } from './systems/Dash'
import { DeathSystem } from './systems/Death'
import { FallDeathSystem } from './systems/FallDeath'
import { GameOverSystem } from './systems/GameOver'
import { InputMovementSystem } from './systems/InputMovement'
import { InputSystem } from './systems/InputSystem'
import { JumpResetSystem } from './systems/JumpReset'
import { MovementValidationSystem } from './systems/MovementValidation'
import { ParallaxRenderSystem } from './systems/ParallaxRender'
import { PlatformSystem } from './systems/Platform'
import { ProjectileSystem } from './systems/ProjectileSystem'
import { RenderSystem } from './systems/RenderSystem'
import { RespawnSystem } from './systems/Respawn'
import { SyncBodyPositionSystem } from './systems/SyncBodyPosition'
import { SyncDirectionSystem } from './systems/SyncDirection'
import { WaypointMovementSystem } from './systems/WaypointMovement'
import { Camera } from './utils/camera'
import * as Factory from './utils/factory'
import * as timer from './utils/timer'

const initializeBlueprints = (manager: Manager) => {
  const blueprints = [
    Factory.createGround,
    Factory.createSlope,
    Factory.createPlatform(200, 300, 400, 20),
    // Factory.createMob(300, 300),
    // Factory.createShieldMob(200, 300),
    Factory.createShootMob(350, 300),
    // Factory.createIcicle(200, 200),
    Factory.createParallax({
      layers: [
        {
          imagePath: 'assets/images/sky.png',
          ratio: 0.1,
        },
        {
          imagePath: 'assets/images/clouds_BG.png',
          ratio: 0.1,
          offsetY: -50,
        },
        {
          imagePath: 'assets/images/mountains.png',
          ratio: 0.125,
          offsetY: -100,
        },
        {
          imagePath: 'assets/images/clouds_MG_3.png',
          ratio: 0.3,
          offsetY: -100,
        },
        {
          imagePath: 'assets/images/clouds_MG_2.png',
          ratio: 0.5,
          offsetY: -125,
        },
        {
          imagePath: 'assets/images/clouds_MG_1.png',
          ratio: 0.6,
          offsetY: -200,
        },
      ],
    }),
  ]

  blueprints.forEach(blueprint => blueprint(manager))
}

export class Game {
  camera: Camera
  gui: Gui
  manager: Manager
  player: Entity
  state: State
  world: World

  constructor() {
    this.world = love.physics.newWorld(0, 9.81 * 30 * 8, true)
    this.world.setCallbacks(
      (a: Fixture, b: Fixture, contact: Contact) => {
        this.manager.beginContact(a, b, contact)
      },
      (a: Fixture, b: Fixture, contact: Contact) => {
        this.manager.endContact(a, b, contact)
      }
    )
    this.camera = new Camera('right')
    this.manager = new Manager(this.world)
    this.manager.addSystems([
      // pre-processors
      new InputSystem(this.manager),
      new GameOverSystem(this.manager),

      // pre-renderers
      new ParallaxRenderSystem(this.manager, this.camera),

      // processors
      new DeathSystem(this.manager),
      new AbilitiesSystem(this.manager),
      new DamageSystem(this.manager),
      new CheckpointSystem(this.manager),
      new ContainerSystem(this.manager),
      new FallDeathSystem(this.manager),
      new InputMovementSystem(this.manager),
      new JumpResetSystem(this.manager),
      new WaypointMovementSystem(this.manager),
      new PlatformSystem(this.manager),
      new ProjectileSystem(this.manager),
      new DashSystem(this.manager),
      new AggressionSystem(this.manager),
      new AttackSystem(this.manager),
      new MovementValidationSystem(this.manager),
      new RespawnSystem(this.manager),

      // post-processors
      new SyncBodyPositionSystem(this.manager),
      new SyncDirectionSystem(this.manager),

      // renderers
      new AnimateSpriteSystem(this.manager),
      new RenderSystem(this.manager),
    ])

    this.player = Factory.createPlayer(this.manager)
    this.gui = new Gui(this.player)

    initializeBlueprints(this.manager)

    this.state = State.PLAYING
  }

  setState = (state: State) => {
    this.state = state
  }

  restart = () => {}

  keypressed = (...args: [KeyConstant, Scancode, boolean]) => {
    this.manager.keypressed(...args)
  }

  keyreleased = (...args: [KeyConstant, Scancode]) => {
    this.manager.keyreleased(...args)
  }

  mousepressed = (...args: [number, number, number, boolean]) => {
    this.manager.mousepressed(...args)
  }

  mousereleased = (...args: [number, number, number, boolean]) => {
    this.manager.mousereleased(...args)
  }

  mousemoved = (x: number, y: number, ...args: any) => {}

  update = (dt: number) => {
    this.gui.update(dt, this)

    if (this.state === State.PLAYING) {
      timer.update(dt)
      this.world.update(dt)
      this.manager.update(dt)

      if (this.player) {
        this.camera.update(this.player)
      }
    }
  }

  draw = () => {
    this.camera.set()
    this.manager.draw()
    this.camera.clear()

    if (this.player) {
      this.gui.draw(this)
    }
  }
}

File: archive/client/src/game/game.lua
local Manager = require 'src.ecs.manager'

local Factory = require 'src.game.factory'
local State = require 'src.game.state'
local HUD = require 'src.game.hud.hud'
local Camera = require 'src.game.utils.camera'
local Ability = require 'src.game.systems.ability'
local Aggression = require 'src.game.systems.aggression'
local AnimateSprite = require 'src.game.systems.animatesprite'
local Attack = require 'src.game.systems.attack'
local Checkpoint = require 'src.game.systems.checkpoint'
local Container = require 'src.game.systems.container'
local Damage = require 'src.game.systems.damage'
local Dash = require 'src.game.systems.dash'
local Death = require 'src.game.systems.death'
local FallDeath = require 'src.game.systems.falldeath'
local FixtureRender = require 'src.game.systems.fixturerender'
local GameOver = require 'src.game.systems.gameover'
local Input = require 'src.game.systems.input'
local InputMovement = require 'src.game.systems.inputmovement'
local JumpReset = require 'src.game.systems.jumpreset'
local Logger = require 'src.game.systems.logger'
local Projectile = require 'src.game.systems.projectile'
local Platform = require 'src.game.systems.platform'
local Respawn = require 'src.game.systems.respawn'
local SetCurrentAnimation = require 'src.game.systems.setcurrentanimation'
local SpriteRender = require 'src.game.systems.spriterender'
local SyncBodyPosition = require 'src.game.systems.syncbodyposition'
local Trigger = require 'src.game.systems.trigger'
local WaveMovement = require 'src.game.systems.wavemovement'
local WaypointMovement = require 'src.game.systems.waypointmovement'

local Game = {}

local function initEntities(factory)
  factory.create(factory.ground())
  factory.create(factory.slope())
  factory.create(factory.platform(400, 200))
  factory.create(factory.crate(100))
  factory.create(factory.crate(164))
  factory.create(factory.icicle(300, 200))
  factory.create(factory.saw(-800, 200))
  factory.create(factory.saw2(-600, 200))
  factory.create(factory.saw3(-400, 200))
  factory.create(factory.checkpoint(1, 500))
  factory.create(factory.checkpoint(2, 700, 200))
  factory.create(
    factory.trigger(
      1500,
      200,
      100,
      100,
      'spawn',
      function()
        factory.create(factory.snowball(2400, 100))
      end
    )
  )
  -- factory.create(factory.wall(300))
  -- factory.create(factory.mob(100, 100))
  -- factory.create(factory.taserMob(100, 100))
  factory.create(factory.slashMob(100, 100))
  -- factory.create(factory.stabMob(100, 100))
  -- factory.create(factory.shootMob(100, 100))
  -- factory.create(factory.shieldMob(100, 100))

  return factory.create(factory.player(-200))
end

function Game:new()
  local world = love.physics.newWorld(0, 9.81, true)
  local manager = Manager:new(world)
  local factory = Factory(world, manager)
  local hud = HUD.new()
  local camera = Camera.new()
  local game = {
    state = State.PLAYING,
    hud = hud,
    camera = camera,
    manager = manager,
    world = world
  }

  love.physics.setMeter(256)

  world.setCallbacks(
    world,
    function(a, b, contact)
      game:beginContact(a, b, contact)
    end,
    function(a, b, contact)
      game:endContact(a, b, contact)
    end
  )

  manager:addSystem(GameOver)
  manager:addSystem(SyncBodyPosition)
  manager:addSystem(Respawn)
  manager:addSystem(Input)
  manager:addSystem(InputMovement)
  manager:addSystem(JumpReset)
  manager:addSystem(WaypointMovement)
  manager:addSystem(WaveMovement)
  manager:addSystem(Aggression)
  manager:addSystem(Attack)
  manager:addSystem(Ability)
  manager:addSystem(Dash)
  manager:addSystem(Damage)
  manager:addSystem(Death)
  manager:addSystem(FallDeath)
  manager:addSystem(Container)
  manager:addSystem(Trigger)
  manager:addSystem(Checkpoint)
  manager:addSystem(Projectile)
  manager:addSystem(Platform)
  manager:addSystem(FixtureRender)
  manager:addSystem(SpriteRender)
  manager:addSystem(SetCurrentAnimation)
  manager:addSystem(AnimateSprite)
  manager:addSystem(Logger)

  manager:setFactory(factory)

  local player = initEntities(factory)

  function game:setState(state)
    self.state = state
  end

  function game:restart()
    love.load()
  end

  function game:keyboard(key, scancode, isrepeat, ispressed)
    self.manager:keyboard(key, scancode, isrepeat, ispressed)
  end

  function game:mouse(x, y, button, istouch, presses)
    self.manager:mouse(x, y, button, istouch, presses)
  end

  function game:beginContact(a, b, contact)
    self.manager:beginContact(a, b, contact)
  end

  function game:endContact(a, b, contact)
    self.manager:endContact(a, b, contact)
  end

  function game:update(dt)
    if self.state == State.PLAYING then
      self.world:update(dt)
      self.manager:update(dt)

      if (player) then
        self.camera:update(dt, player)
      end
    end

    self.hud:update(dt, player, self)
  end

  function game:draw()
    self.camera:set()
    self.manager:draw()
    self.camera:clear()

    if player then
      self.hud:draw(player, self)
    end
  end

  return game
end

return Game
